### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by vinjana**

### CODE REVIEW

File: SophiaWorkflow.iml [hunk 1]
```diff
@@ -24,5 +24,6 @@
     <orderEntry type="module" module-name="RoddyToolLib.main" />
     <orderEntry type="module" module-name="DefaultPlugin_1.2.2" />
     <orderEntry type="module" module-name="PluginBase_1.2.1" />
+    <orderEntry type="library" name="scala-sdk-2.13.3" level="application" />
```
Comment (vinjana) [inline SophiaWorkflow.iml:27]: This is only for the comparison script.

File: compareSVsTSVs [hunk 2]
```diff
@@ -0,0 +1,309 @@
+#!/usr/bin/env scala
```
Comment (vinjana) [inline compareSVsTSVs:1]: You don't have to review this. This is only for comparing to result directories, to check for changes in the output.

File: buildinfo.txt [hunk 3]
```diff
@@ -1,4 +1,4 @@
-dependson=COWorkflowsBasePlugin:1.4.0
+dependson=COWorkflowsBasePlugin:1.4.1
```
Comment (vinjana) [inline buildinfo.txt:1]: Only robustness-changes (related to checks for IO-errors). This version is already in use in recent Indel and should be unproblematic.

File: resources/analysisTools/sophiaworkflow/fusionCandidates.py [hunk 4]
```diff
@@ -29,59 +29,93 @@
...
 import itertools
-from sys import stderr
+import sys
+
 
 MAXDISTANCE = 2000000
 
 
-def natural_sort(l):
+def natural_string_sort(strings: list) -> list:
```
Comment (vinjana) [inline resources/analysisTools/sophiaworkflow/fusionCandidates.py:38]: Note that the type annotation `: list` in Python 3 is not binding, which means Python is not checking it, but it is helpful in IntelliJ, where type mismatches are checked and highlighted!

File: compareSVsTSVs [hunk 5]
```diff
@@ -152,10 +147,10 @@ class Line(val fields: List[String]) {
           requireConsistentNonRawGeneSets(idx, this.fields(idx), that.fields(idx))
 
         } else if (fusionExpressionColumnIndices.contains(idx)) {
-          // For now the fusion-expression-valued columns of fusion pairs are ignored. Testing them like here
-          // (with much less control over the input than in a unit-test) would mean re-implementing the
-          // feature (gene-name extraction).
-          throw new NotImplementedError()
```
Comment (jpalbrecht) [inline compareSVsTSVs]: why is this deleted? Is the comment reason enough for not having any code at this position?

File: SophiaWorkflow.iml [hunk 6]
```diff
@@ -19,8 +19,10 @@
     <orderEntry type="sourceFolder" forTests="false" />
     <orderEntry type="library" name="Gradle: org.codehaus.groovy:groovy-all:2.4.9" level="project" />
     <orderEntry type="library" name="Python 3.7 (base) interpreter library" level="application" />
-    <orderEntry type="module" module-name="Roddy" />
     <orderEntry type="module" module-name="COWorkflowsBasePlugin_1.4.1" />
+    <orderEntry type="module" module-name="Roddy.main" />
```
Comment (jpalbrecht) [inline SophiaWorkflow.iml]: hope this is ok - cannot really say if or if not

File: README.md [hunk 7]
```diff
@@ -110,12 +145,16 @@ the processed gene identifiers columns, results from <2.2.2 and >=2.2.2 cannot b
...
+
+  * Added comparison script `compareSvsTSVs` for `svs*.tsv` files.
+  * Added basic check to shield against yet unfixed potential unnoticable IO-errors in sophia binary (see [Bitbucket PR](https://bitbucket.org/utoprak/sophia/pull-requests/2/safer-file-io/diff))
+  * Refactorings and documentation to improve code clarity
+
 * 2.2.2
 
   * Bugfix in `intersectionCollapsing.py`: Produced columns with `set()` output and lines were put out instead of withheld, resulting in alignment errors in subsequent unchecked paste commands. Bug was fixed, script refactored and Python 3 type hints employed (not enforced by Python, though).
   * Added `safe_paste.py` to shield against similar bugs in the future
-  * Added basic check to shield against yet unfixed potential unnoticable IO-errors in sophia binary (see [Bitbucket PR](https://bitbucket.org/utoprak/sophia/pull-requests/2/safer-file-io/diff))
```
Comment (jpalbrecht) [inline README.md:83]: is this feature in 2.2.2??? If, i would leave this line here, although it was not intended for this realease.

File: compareSvsTSVs [hunk 8]
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env scala
```
Comment (jpalbrecht) [inline compareSvsTSVs]: took us 5 mins to figure out: This is a different name than the file created in commit "Added comparison script for Sophia result directories. No complete co… " <- not a good idea in our opinion

File: resources/analysisTools/sophiaworkflow/fusionCandidates.py [hunk 9]
```diff
@@ -29,59 +29,90 @@
...
+        # `fusion` variable is added to the output list.
+        fusionID = '-'.join(natural_string_sort(fusion))
+        if fusionID not in sortedFusions:
+            sortedFusions.add(fusionID)
+            filteredFusionList.append(fusion)
+    return filteredFusionList
+
+
+def getDirectFusionStr(gene1Raw: str, gene2Raw: str) -> str:
+    """Given two strings of comma-separated extended gene names, produce a string representing direct fusions. The
```
Comment (jpalbrecht) [inline resources/analysisTools/sophiaworkflow/fusionCandidates.py:95]: nice! Help strings help!

File: resources/analysisTools/sophiaworkflow/fusionCandidates.py [hunk 10]
```diff
@@ -91,79 +122,57 @@ def getDirectFusion(gene1Raw, gene2Raw):
...
+        gene2Raw =                     extractIdListFromListOfIdSemicolonElems(lineChunks[30])
+        gene2RawCancer =               extractIdListFromListOfIdSemicolonElems(lineChunks[31])
+        gene2NearestUpstreamRaw =      distanceFilter(extractIdListFromListOfIdSemicolonElems(lineChunks[32]),
+                                                      distance=int(lineChunks[33]))
+        gene2NearestUpstreamCancer =   distanceFilter(extractIdListFromListOfIdSemicolonElems(lineChunks[34]),
+                                                      distance=int(lineChunks[35]))
+        gene2NearestDownstreamRaw =    distanceFilter(extractIdListFromListOfIdSemicolonElems(lineChunks[36]),
+                                                      distance=int(lineChunks[37]))
+        gene2NearestDownstreamCancer = distanceFilter(extractIdListFromListOfIdSemicolonElems(lineChunks[38]),
+                                                      distance=int(lineChunks[39]))
```
Comment (jpalbrecht) [inline resources/analysisTools/sophiaworkflow/fusionCandidates.py]: finally, thanks for removing the eye-cancer!

File: resources/analysisTools/sophiaworkflow/fusionCandidates.py [hunk 11]
```diff
@@ -127,145 +128,151 @@ def extractIdListFromListOfIdSemicolonElems(string: str) -> str:
...
-        print(line.rstrip(),
-              directFusionCandidates, directFusionCandidatesBothCancer,
-              indirectFusionCandidatesLeftCancerRightAny, indirectFusionCandidatesRightCancerLeftAny,
-              indirectFusionCandidatesAny, sep='\t')
+            print(line.rstrip(),
+                  directFusionCandidates, directFusionCandidatesBothCancer,
+                  indirectFusionCandidatesLeftCancerRightAny, indirectFusionCandidatesRightCancerLeftAny,
+                  indirectFusionCandidatesAny, sep='\t')
+except Exception as e:
+    print("Error while processing line {}: {}".format(line_no, e), file=sys.stderr)
```
Comment (jpalbrecht) [inline resources/analysisTools/sophiaworkflow/fusionCandidates.py:225]: :+1:

File: resources/analysisTools/sophiaworkflow/fusionCandidates.py [hunk 12]
```diff
@@ -177,102 +199,28 @@ def distanceFilter(value: str, distanceStr: str, max_distance: int = MAXDISTANCE
...
-                        leftComponent += "~" + gene1NearestUpstreamCancer
-                    leftComponent += "/"
-                    if gene1NearestDownstreamCancer not in {"", "."}:
-                        leftComponent += "~" + gene1NearestDownstreamCancer
-                else:
-                    leftComponent = "(TRUNC)"
-            else:
-                leftComponent = gene1RawCancer
+            indirectFusionCandidatesLeftCancerRightAny = \
+                getIndirectFusionStr(gene1RawCancer, gene1NearestUpstreamCancer, gene1NearestDownstreamCancer,
```
Comment (jpalbrecht) [inline resources/analysisTools/sophiaworkflow/fusionCandidates.py:203]: :1st_place_medal:

File: resources/analysisTools/sophiaworkflow/fusionCandidates.py [hunk 13]
```diff
@@ -177,102 +199,28 @@ def distanceFilter(value: str, distanceStr: str, max_distance: int = MAXDISTANCE
...
-            else:
-                rightComponent = firstRawGeneName(gene2Raw)
-            if leftComponent != rightComponent:
-                indirectFusionCandidatesAny = leftComponent + "-" + rightComponent
-            else:
-                indirectFusionCandidatesAny = "."
-
-            directFusionCandidates = getDirectFusionStr(gene1Raw, gene2Raw)
-            directFusionCandidatesBothCancer = getDirectFusionStr(gene1RawCancer, gene2RawCancer)
+            directFusionCandidates = \
```
Comment (jpalbrecht) [inline resources/analysisTools/sophiaworkflow/fusionCandidates.py:214]: :+1: :+1: :+1:

File: compareSVsTSVs (hunk 2 above)
Comment (jpalbrecht) [reply compareSVsTSVs:1]: see commit "Added compareSvsTSVs. Added start of svs-TSV column descriptions do t "

**Review by jpalbrecht — APPROVED** [review]

File: compareSVsTSVs (hunk 2 above)
Comment (vinjana) [reply compareSVsTSVs:1]: Note sure about what you mean by '"co..." <- no good option in our opinion'. Are you referring to the script-name? Or are you referring to the commit-description? (indeed a bit "unlucky" if truncated in Github).

**Review by vinjana — COMMENTED** [review]
Thanks for the review!
